### PR TITLE
feat(api): add endpoint for creating users

### PR DESCRIPTION
**Pull request by arminpatel**
+ `view` : generic CreateAPI view
+ `Post` request at `/users/` for creating users
+ ` Serializers` 
  + Simple Model Serializers for `Faculty`, `Student` and `Staff`
  + These three serializers are nested inside `UserSerializer` and their `required` attribute is set to `False`
  + The `create` method then extracts data of each Serializer(if present) and the remaining data is used to create the user. And depending on the extracted data, the student, faculty and staff instances are created.
  +  The problem I was facing in this implementation was that a User could be only one of staff, faculty and student. I've have handled it by assuming only one or neither of them would be passed to the api. Let me know if the design could have been better.
+ `testing`: I have added some test to cover different scenarios like adding base User, Student, Faculty etc, and what should happen when invalid emails are provided

### CODE REVIEW

**Comment by Samy-33** [comment]
Left some comments on https://github.com/BitByte-TPC/gymkhana/pull/164.

Closing a PR if you mess something up in git and reopening a fresh PR is not good for multiple reasons:
1. You have to do this toiling task every time you mess up something in `git`.
2. Reviews done in the previous PR is kind of lost.
3. Git provides ways to do it in an easier way, so you should utilise that. 
    - [git reflog](https://www.atlassian.com/git/tutorials/rewriting-history/git-reflog#:~:text=Git%20keeps%20track%20of%20updates,git%20reset)
    - [git rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

**Comment by arminpatel** [comment]
> Left some comments on #164.
> 
> Closing a PR if you mess something up in git and reopening a fresh PR is not good for multiple reasons:
> 
> 1. You have to do this toiling task every time you mess up something in `git`.
> 2. Reviews done in the previous PR is kind of lost.
> 3. Git provides easier ways to fix the mess up, so you should utilise that.
>    
>    * [git reflog](https://www.atlassian.com/git/tutorials/rewriting-history/git-reflog#:~:text=Git%20keeps%20track%20of%20updates,git%20reset)
>    * [git rebase](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)

Ok, will mind from next time. So should I fix and reopen #164 or keep this one?

**Comment by arminpatel** [comment]
Tried to use pytest parametrize and fixtures but was thrown with the following error,

Screenshot 2022-07-16 at 8 28 13 AM

[my code](https://p.ip.fi/b6Ih)

Pytest fixtures and parametrize do not support django unittest.TestCase subclass
[reference](https://docs.pytest.org/en/latest/how-to/unittest.html#pytest-features-in-unittest-testcase-subclasses)
So is there any other way?

**Comment by Samy-33** [comment]
Have you tried placing `@parameterize` decorator below the `@django_db` decorator?

As decorator's order matters in python, it might be the issue.

- Don't use variables in the parameters list, just use the values. Readers have to go to the variable definition to know the value.

-`invalidEmailProvided`, `throwsBadRequest`

**Comment by arminpatel** [comment]
> Have you tried placing `@parameterize` decorator below the `@django_db` decorator?
> 
> As decorator's order matters in python, it might be the issue.

Still getting the same error. [my code](https://p.ip.fi/JTks)

**Comment by Samy-33** [comment]
@arminpatel can you push your code, I will take a look.

**Comment by Samy-33** [comment]
@arminpatel let's use function based tests for `CreateUserView`

```python
@pytest.mark.parametrize('invalid_email', ['', 'someemail'])
@pytest.mark.django_db
def testUsersListView_invalidEmailsProvided_returnsHttp400(invalid_email):
    # given
    user = User.objects.create(
        email=primaryTestUserMail, first_name='Test', last_name='User')
    client = APIClient()
    client.force_authenticate(user)

    # when
    response1 = client.post(userCreationEndpoint, {'email': invalid_email},
                            format='json')

    # then
    assert response1.status_code == status.HTTP_400_BAD_REQUEST
```

**Comment by arminpatel** [comment]
> @arminpatel let's use function based tests for `CreateUserView`
> We need to refactor all the tests to be in functional format instead of class based.

ok, I will make the changes

**Comment by arminpatel** [comment]
@Samy-33 Made the changes.

File: api/api/accounts/tests.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                            'user_type': 'Staff',
+                            'staff': {
+                                'department': 'ECE',
+                                'designation': 'z'
+                                }
+                            })
+
+    # then
+    assert response.status_code == status.HTTP_201_CREATED
+    assert response.data['email'] == test_user_email
```
Comment (Samy-33) [inline api/api/accounts/tests.py:158]: redundant check?

File: api/api/accounts/tests.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                             'user_type': 'Student',
+                             'picture_url': '',
+                             'student': None,
+                             'faculty': None,
+                             'staff': None
+                             }
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('invalid_mail', ['test@user.com', '', 'testusersom'])
```
Comment (Samy-33) [inline api/api/accounts/tests.py:60]: Let's add comment for each invalid email as to why it's invalid.

```python
@pytest.mark.parametrize('invalid_mail', [
    'test@user.com', # admin user 
    '', # empty email
    'testusersom' # invalid email format
])
```

File: api/api/accounts/views.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from rest_framework import generics
+from rest_framework.permissions import IsAdminUser
+
+from .models import User
```
Comment (Samy-33) [inline api/api/accounts/views.py:4]: use `get_user_model()`

File: api/api/accounts/tests.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+user_creation_endpoint = '/users/'
+primary_test_user_mail = 'test@user.com'
+
+
+@pytest.mark.django_db
+def testCreateUserView_testPermission_nonAdminRequest_throwsForbidden():
+    # given
+    user = User.objects.create(
+        email=primary_test_user_mail, first_name='Test', last_name='User')
+    client = APIClient()
```
Comment (Samy-33) [inline api/api/accounts/tests.py:18]: nit: we can use `pytest.fixture` to create repetitive objects once.

```python
@pytest.mark.django_db
@pytest.fixture(scope='function')
def admin_user(django_user_model):
    return django_user_model.objects.create(email=primary_test_user_mail, first_name='Test', last_name='User', is_staff=True)

@pytest.fixture(scope='function')
def client():
    return APIClient()
```

And then use it like:
```python
@pytest.mark.django_db
def testCreateUserView_testPermission_nonAdminRequest_throwsForbidden(admin_user, client):
    ...
```

`pytest_django` also allows you to use `django_user_model` fixture to get `User` model: [Reference](https://pytest-django.readthedocs.io/en/latest/helpers.html#django-user-model)

File: api/api/accounts/tests.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                            'student': {
+                                'roll_no': "21abc000",
+                                'batch': 2025,
+                                'department': "CSE",
+                                'hostel_address': "H3-Z000"
+                                }
+                            })
+
+    # then
+    print(response.data)
```
Comment (Samy-33) [inline api/api/accounts/tests.py:100]: Remove print

**Review by Samy-33 — CHANGES_REQUESTED** [review]
Overall looks good, few minor improvements.

File: api/api/accounts/tests.py (hunk 1 above)
Comment (arminpatel) [reply api/api/accounts/tests.py:158]: Yeah, thought more checks wouldn't hurt 😅 . I will remove them.

**Review by Samy-33 — APPROVED** [review]
Looks great. Good work @arminpatel

**Pull request merged**